### ISSUE TITLE
Embrace more types on ObjectBuilder? Enable JSON-Schema linkage

### ISSUE THREAD

**Issue by rmannibucau** [eclipse-ee4j/jsonp#103]
JsonObjectBuilder is a very convenient API and has the advantage in current form to not loose immediately the format of the underlying data (BigDecimal/BigInteger, long, ....). However I think it should cover what developpers consider as "primitives". To be accurate I think we can use json schema specification as a basis.

What is missing IMO is:

- date-time (can use java 8 types and formatters as "normalizer" when it goes to JsonObject)
- date (can use java 8 types and formatters as "normalizer" when it goes to JsonObject)
- time (can use java 8 types and formatters as "normalizer" when it goes to JsonObject)
- email, hostname (think these ones can be collapsed with string since no other format uses this information I think)
- ipaddress (java.net)
- resource identifier (URI)
- regex (this one is to refine since I'm not sure a Pattern can easily and accurately be converted to a string all the times)
- ...

The most important (and urgent) item is likely about the dates but longer term it would be very good to have an API allowing users or implementations to generate a schema from the builder API transparently (just decorating the default implementations).

**Comment by keilw** [comment]
I think there was a similar question about Date/Time. 
These types are not in the JSON standard (RFC) so while JSON-B provides out of the box mapping to certain types it does not seem covered by the standard. @m0mus WDYT about this?

**Comment by rmannibucau** [comment]
@keilw issue is not in jsongenerator or components like that but jsonobjectbuilder which is java first so jsonobjectbuilder should provide a stronger typing than json and lower level components IMHO.

**Comment by keilw** [comment]
@rmannibucau I'm talking about `JsonObjectBuilder` which is only creates a `JsonObject` right now. 
Do you suggest to add methods like
` JsonObjectBuilder add(String name, LocalDateTime value);` or similar?
 Not sure, if this fits in a lower level API, but @m0mus has to say, how he feels it should evolve if any beyond the JSON specification.

**Comment by rmannibucau** [comment]
Well, the builder is an awesome api to

1. Enrich the generation with meta (jsonschema)
2. Create a JsonObject backed with another serialization like avro or protobuf

Only issue is it misses some type info so would be great to fill the gap to avoid another api on top of it

**Comment by keilw** [comment]
The thing is there IS another API on top of that [JSON-B](https://github.com/eclipse-ee4j/jsonb-api). I leave it to @m0mus what aspects to add or cover here, some I have the feeling belong to JSON-B, and it is up to users if they are happy with low-level functionality of the strict JSON standard or want a richer type-system.

**Comment by rmannibucau** [comment]
@keilw no, JSON-B doesnt overlap this use case at all, the opposite is true, enhacing jsonp to support that would benefit json-b but the opposite will not.

**Comment by keilw** [comment]
@m0mus WDYT, does it belong here or rather JSON-B?

**Comment by m0mus** [comment]
I'm fine to add it to JSONP.
